File: internal/rules/formatter.go
package rules

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/kx0101/replayer/internal/cli"
)

func FormatRuleResult(result *RuleEvaluationResult) string {
	var sb strings.Builder

	sb.WriteString("\n")
	sb.WriteString("═══════════════════════════════════════════════════════\n")
	sb.WriteString("            REGRESSION RULES EVALUATION\n")
	sb.WriteString("═══════════════════════════════════════════════════════\n")
	sb.WriteString("\n")

	if result.Passed {
		sb.WriteString("PASSED - All rules satisfied\n")
		sb.WriteString("═══════════════════════════════════════════════════════\n")

		return sb.String()
	}

	sb.WriteString(fmt.Sprintf("FAILED - %d rule violation(s) detected\n", len(result.Failures)))
	sb.WriteString("\n")

	for i, failure := range result.Failures {
		sb.WriteString("───────────────────────────────────────────────────────\n")
		sb.WriteString(fmt.Sprintf("Failure #%d\n", i+1))
		sb.WriteString("───────────────────────────────────────────────────────\n")
		sb.WriteString(fmt.Sprintf("Rule:    %s\n", failure.Rule))
		sb.WriteString(fmt.Sprintf("Scope:   %s\n", failure.Scope))
		sb.WriteString(fmt.Sprintf("Message: %s\n", failure.Message))

		if len(failure.Details) > 0 {
			sb.WriteString("\nDetails:\n")

			for key, value := range failure.Details {
				sb.WriteString(fmt.Sprintf("  %s: %v\n", key, value))
			}
		}
		sb.WriteString("\n")
	}

	sb.WriteString("═══════════════════════════════════════════════════════\n")

	return sb.String()
}

func FormatRuleResultJSON(result *RuleEvaluationResult) (string, error) {
	data, err := json.MarshalIndent(result, "", "  ")
	if err != nil {
		return "", fmt.Errorf("failed to marshal result: %w", err)
	}

	return string(data), nil
}

func GetExitCode(result *RuleEvaluationResult) cli.ExitCode {
	if result.Passed {
		return cli.ExitOK
	}

	return cli.ExitRules
}

File: crates/replayer-cli/src/rules/formatter.rs
use std::fmt::Write;

use replayer_core::ExitCode;

use super::types::RuleEvaluationResult;

pub fn format_rule_result(result: &RuleEvaluationResult) -> String {
    let mut sb = String::new();

    writeln!(sb).unwrap();
    writeln!(
        sb,
        "═══════════════════════════════════════════════════════"
    )
    .unwrap();
    writeln!(sb, "            REGRESSION RULES EVALUATION").unwrap();
    writeln!(
        sb,
        "═══════════════════════════════════════════════════════"
    )
    .unwrap();
    writeln!(sb).unwrap();

    if result.passed {
        writeln!(sb, "PASSED - All rules satisfied").unwrap();
        writeln!(
            sb,
            "═══════════════════════════════════════════════════════"
        )
        .unwrap();
        return sb;
    }

    writeln!(
        sb,
        "FAILED - {} rule violation(s) detected",
        result.failures.len()
    )
    .unwrap();
    writeln!(sb).unwrap();

    for (i, failure) in result.failures.iter().enumerate() {
        writeln!(
            sb,
            "───────────────────────────────────────────────────────"
        )
        .unwrap();
        writeln!(sb, "Failure #{}", i + 1).unwrap();
        writeln!(
            sb,
            "───────────────────────────────────────────────────────"
        )
        .unwrap();
        writeln!(sb, "Rule:    {}", failure.rule).unwrap();
        writeln!(sb, "Scope:   {}", failure.scope).unwrap();
        writeln!(sb, "Message: {}", failure.message).unwrap();

        if !failure.details.is_empty() {
            writeln!(sb).unwrap();
            writeln!(sb, "Details:").unwrap();
            for (key, value) in &failure.details {
                writeln!(sb, "  {}: {}", key, value).unwrap();
            }
        }
        writeln!(sb).unwrap();
    }

    writeln!(
        sb,
        "═══════════════════════════════════════════════════════"
    )
    .unwrap();

    sb
}

pub fn format_rule_result_json(result: &RuleEvaluationResult) -> anyhow::Result<String> {
    let json = serde_json::to_string_pretty(result)
        .map_err(|e| anyhow::anyhow!("failed to marshal result: {}", e))?;
    Ok(json)
}

pub fn get_exit_code(result: &RuleEvaluationResult) -> ExitCode {
    if result.passed {
        ExitCode::Ok
    } else {
        ExitCode::Rules
    }
}
